### PR TITLE
Feature/dec macro float literals

### PR DESCRIPTION
**Pull request by dhedey**
## Summary
* Improves error codes from parsing `Decimal` and `PreciseDecimal`
* Adds support for float literals in `dec` and `pdec`, ie you can now do `dec!(123.112321)`
* Removes support for `dec!(true)` and `dec!(false)` as this is potentially confusing

### For dApp Developers
You can remove your quotes/strings in `dec!` and `pdec!` if you wish.
For example, `dec!("123.452")` can become `dec!(123.452)`.

### CODE REVIEW

File: radix-engine-common/src/math/bnum_integer/convert.rs [hunk 1]
```diff
@@ -219,7 +219,14 @@ macro_rules! impl_from_string {
                     fn from_str(val: &str) -> Result<Self, Self::Err> {
                         match <$wrapped>::from_str(val) {
                             Ok(val) => Ok(Self(val)),
-                            Err(_) => Err([<Parse $t Error>]::InvalidDigit),
+                            Err(err) => Err(match err.kind() {
+                                core::num::IntErrorKind::Empty => [<Parse $t Error>]::Empty,
+                                core::num::IntErrorKind::InvalidDigit => [<Parse $t Error>]::InvalidDigit,
+                                core::num::IntErrorKind::PosOverflow => [<Parse $t Error>]::Overflow,
+                                core::num::IntErrorKind::NegOverflow => [<Parse $t Error>]::Overflow,
+                                core::num::IntErrorKind::Zero => unreachable!(),
```
Comment (iamyulong) [inline radix-engine-common/src/math/bnum_integer/convert.rs:227]: Minor: `unreachable!("Zero is only issued for non-zero type")`

File: radix-engine-common/src/math/decimal.rs [hunk 2]
```diff
@@ -740,36 +740,51 @@ impl FromStr for Decimal {
...
         let v: Vec<&str> = s.split('.').collect();
 
-        let mut int = match I192::from_str(v[0]) {
+        let integer_part = match I192::from_str(v[0]) {
             Ok(val) => val,
-            Err(_) => return Err(ParseDecimalError::InvalidDigit),
+            Err(err) => match err {
+                ParseI192Error::NegativeToUnsigned => unreachable!("Not possible to be thrown by from_str"),
+                ParseI192Error::Overflow => return Err(ParseDecimalError::Overflow),
+                ParseI192Error::InvalidLength => unreachable!("Not possible to be thrown by from_str"),
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:750]: Minor: `unreachable!("InvalidLength is only for parsing &[u8], not &str")`

File: radix-engine-common/src/math/decimal.rs [hunk 3]
```diff
@@ -740,36 +740,51 @@ impl FromStr for Decimal {
...
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let tens = I192::TEN;
         let v: Vec<&str> = s.split('.').collect();
 
-        let mut int = match I192::from_str(v[0]) {
+        let integer_part = match I192::from_str(v[0]) {
             Ok(val) => val,
-            Err(_) => return Err(ParseDecimalError::InvalidDigit),
+            Err(err) => match err {
+                ParseI192Error::NegativeToUnsigned => unreachable!("Not possible to be thrown by from_str"),
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:748]: Minor: `unreachable!("NegativeToUnsigned is only for parsing unsigned type, not I192"]`

File: radix-engine-common/src/math/decimal.rs [hunk 4]
```diff
@@ -740,36 +740,51 @@ impl FromStr for Decimal {
...
+                ParseI192Error::Overflow => return Err(ParseDecimalError::Overflow),
+                ParseI192Error::InvalidLength => unreachable!("Not possible to be thrown by from_str"),
+                ParseI192Error::InvalidDigit => return Err(ParseDecimalError::InvalidDigit),
+                ParseI192Error::Empty => I192::ZERO,
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:752]: What is `Empty`? Same for all the other places.

Comment (iamyulong) [reply radix-engine-common/src/math/decimal.rs:752]: As discussed. This is to support `.123` and `123.`

**Review by iamyulong — APPROVED** [review]

File: radix-engine-common/src/math/decimal.rs (hunk 4 above)
Comment (dhedey) [reply radix-engine-common/src/math/decimal.rs]: Although we have now agreed to not allow this. Have tweaked it and added a comment.

**Pull request merged**